libs/execution_context/ebpf_native.c

```diff
@@ -219,20 +229,34 @@ _ebpf_native_clean_up_maps(_In_reads_(map_count) _Frees_ptr_ ebpf_native_map_t*
 }
 
 static void
-_ebpf_native_clean_up_programs(_In_reads_(count_of_programs) ebpf_native_program_t* programs, size_t count_of_programs)
+_ebpf_native_clean_up_program(_In_opt_ _Post_invalid_ ebpf_native_program_t* program, bool close_handle)
 {
-    for (uint32_t i = 0; i < count_of_programs; i++) {
-        if (programs[i].handle != ebpf_handle_invalid) {
+    if (program != NULL) {
+        if (program->handle != ebpf_handle_invalid) {
             ebpf_program_t* program_object = NULL;
             ebpf_assert_success(EBPF_OBJECT_REFERENCE_BY_HANDLE(
-                programs[i].handle, EBPF_OBJECT_PROGRAM, (ebpf_core_object_t**)&program_object));
+                program->handle, EBPF_OBJECT_PROGRAM, (ebpf_core_object_t**)&program_object));
             ebpf_assert_success(ebpf_program_register_for_helper_changes(program_object, NULL, NULL));
             EBPF_OBJECT_RELEASE_REFERENCE((ebpf_core_object_t*)program_object);
-            ebpf_assert_success(ebpf_handle_close(programs[i].handle));
-            programs[i].handle = ebpf_handle_invalid;
+            if (close_handle) {
+                ebpf_assert_success(ebpf_handle_close(program->handle));
+                program->handle = ebpf_handle_invalid;
+            }
         }
-        ebpf_free(programs[i].addresses_changed_callback_context);
-        programs[i].addresses_changed_callback_context = NULL;
+        ebpf_free(program->addresses_changed_callback_context);
+        program->addresses_changed_callback_context = NULL;
+        ebpf_free(program->runtime_context.helper_data);
+        ebpf_free(program->runtime_context.map_data);
+        ebpf_free(program);
+    }
+}
+
+static void
+_ebpf_native_clean_up_programs(
+    _In_reads_(count_of_programs) ebpf_native_program_t** programs, size_t count_of_programs, bool close_handles)
```

Comment:
Just double checking - are we guaranteed that programs is always non-null?

Reply:
Seems like it.  If it isn't, the SAL annotation here should cause the compiler to complain.